From 61195df50e2ced675262fbafe0b672f12b832b91 Mon Sep 17 00:00:00 2001
From: Dudemanguy <random342@airmail.cc>
Date: Wed, 20 Mar 2024 17:17:39 -0500
Subject: [PATCH] build: Fix build without libsystemd

The graphical remote login feature requires systemd, but this can be
made optional so the usual remote assistance sessions can still work.

Closes: https://gitlab.gnome.org/GNOME/gnome-remote-desktop/-/issues/190
---
 config.h.meson   |  3 +++
 meson.build      |  1 +
 src/grd-daemon.c | 22 ++++++++++++++++------
 src/meson.build  | 22 +++++++++++++++-------
 4 files changed, 35 insertions(+), 13 deletions(-)

diff --git a/config.h.meson b/config.h.meson
index 3dff24f1d..9942c6c69 100644
--- a/config.h.meson
+++ b/config.h.meson
@@ -6,6 +6,9 @@
 /* The prefix for our gettext translation domains. */
 #mesondefine GETTEXT_PACKAGE
 
+/* Defined if LIBSYSTEMD is supported */
+#mesondefine HAVE_LIBSYSTEMD
+
 /* Defined if RDP backend enabled */
 #mesondefine HAVE_RDP
 
diff --git a/meson.build b/meson.build
index 54adb1c8b..cdfec4f1a 100644
--- a/meson.build
+++ b/meson.build
@@ -94,6 +94,7 @@ cdata = configuration_data()
 cdata.set_quoted('GETTEXT_PACKAGE', 'gnome-remote-desktop')
 cdata.set_quoted('VERSION', meson.project_version())
 
+cdata.set('HAVE_LIBSYSTEMD', libsystemd_dep.found())
 cdata.set('HAVE_RDP', have_rdp)
 cdata.set('HAVE_VNC', have_vnc)
 
diff --git a/src/grd-daemon.c b/src/grd-daemon.c
index d59bd81d8..89aecf711 100644
--- a/src/grd-daemon.c
+++ b/src/grd-daemon.c
@@ -36,8 +36,6 @@
 #include <stdlib.h>
 
 #include "grd-context.h"
-#include "grd-daemon-handover.h"
-#include "grd-daemon-system.h"
 #include "grd-daemon-user.h"
 #include "grd-dbus-mutter-remote-desktop.h"
 #include "grd-dbus-remote-desktop.h"
@@ -48,6 +46,11 @@
 #include "grd-utils.h"
 #include "grd-vnc-server.h"
 
+#ifdef HAVE_LIBSYSTEMD
+#include "grd-daemon-handover.h"
+#include "grd-daemon-system.h"
+#endif /* HAVE_LIBSYSTEMD */
+
 #define RDP_SERVER_RESTART_DELAY_MS 3000
 #define RDP_SERVER_USER_CERT_SUBDIR "certificates"
 #define RDP_MAX_CERTIFICATE_FILE_SIZE_BYTES (50 * 1024)
@@ -983,11 +986,11 @@ get_daemon_dbus_connection (GrdDaemon *daemon)
   g_autoptr (GDBusConnection) connection = NULL;
   g_autoptr (GError) error = NULL;
 
-#ifdef HAVE_RDP
+#if defined(HAVE_RDP) && defined(HAVE_LIBSYSTEMD)
   if (GRD_IS_DAEMON_SYSTEM (daemon))
     connection = g_bus_get_sync (G_BUS_TYPE_SYSTEM, NULL, &error);
   else
-#endif /* HAVE_RDP */
+#endif /* HAVE_RDP && HAVE_LIBSYSTEMD */
     connection = g_bus_get_sync (G_BUS_TYPE_SESSION, NULL, &error);
 
   if (!connection)
@@ -1240,12 +1243,12 @@ main (int argc, char **argv)
       "Print version", NULL },
     { "headless", 0, 0, G_OPTION_ARG_NONE, &headless,
       "Run in headless mode", NULL },
-#ifdef HAVE_RDP
+#if defined(HAVE_RDP) && defined(HAVE_LIBSYSTEMD)
     { "system", 0, 0, G_OPTION_ARG_NONE, &system,
       "Run in headless mode as a system g-r-d service", NULL },
     { "handover", 0, 0, G_OPTION_ARG_NONE, &handover,
       "Run in headless mode taking a connection from system g-r-d service", NULL },
-#endif
+#endif /* HAVE_RDP && HAVE_LIBSYSTEMD */
     { "rdp-port", 0, 0, G_OPTION_ARG_INT, &rdp_port,
       "RDP port", NULL },
     { "vnc-port", 0, 0, G_OPTION_ARG_INT, &vnc_port,
@@ -1295,12 +1298,19 @@ main (int argc, char **argv)
     case GRD_RUNTIME_MODE_HEADLESS:
       daemon = GRD_DAEMON (grd_daemon_user_new (runtime_mode, &error));
       break;
+#ifdef HAVE_LIBSYSTEMD
     case GRD_RUNTIME_MODE_SYSTEM:
       daemon = GRD_DAEMON (grd_daemon_system_new (&error));
       break;
     case GRD_RUNTIME_MODE_HANDOVER:
       daemon = GRD_DAEMON (grd_daemon_handover_new (&error));
       break;
+#else
+    case GRD_RUNTIME_MODE_SYSTEM:
+    case GRD_RUNTIME_MODE_HANDOVER:
+      g_assert_not_reached ();
+      break;
+#endif /* HAVE_LIBSYSTEMD */
     }
 
   if (!daemon)
diff --git a/src/meson.build b/src/meson.build
index 5f01b4098..248117933 100644
--- a/src/meson.build
+++ b/src/meson.build
@@ -16,7 +16,6 @@ deps = [
   pipewire_dep,
   libei_dep,
   libnotify_dep,
-  libsystemd_dep,
   epoxy_dep,
   drm_dep,
   xkbcommon_dep,
@@ -82,12 +81,6 @@ if have_rdp
   daemon_sources += files([
     'grd-clipboard-rdp.c',
     'grd-clipboard-rdp.h',
-    'grd-daemon-handover.c',
-    'grd-daemon-handover.h',
-    'grd-daemon-system.c',
-    'grd-daemon-system.h',
-    'grd-daemon-utils.c',
-    'grd-daemon-utils.h',
     'grd-hwaccel-nvidia.c',
     'grd-hwaccel-nvidia.h',
     'grd-rdp-audio-input.c',
@@ -164,6 +157,21 @@ if have_rdp
     'grd-session-rdp.h',
   ])
 
+  if libsystemd_dep.found()
+    daemon_sources += files([
+      'grd-daemon-handover.c',
+      'grd-daemon-handover.h',
+      'grd-daemon-system.c',
+      'grd-daemon-system.h',
+      'grd-daemon-utils.c',
+      'grd-daemon-utils.h',
+    ])
+
+    deps += [
+      libsystemd_dep,
+    ]
+  endif
+
   deps += [
     cuda_dep,
     dl_dep,
-- 
GitLab

